### PR TITLE
Remove empty root when possible Version 2

### PR DESCRIPTION
**Pull request by dboehm-avalabs**
Currently the MerkleDB always has the node with the nil key as its root. This isn't required for the trie structure whenever the root has no value and only a single child. In those cases, switch to using that child as the root of the trie.

Alternative to https://github.com/ava-labs/avalanchego/pull/2106

### CODE REVIEW

File: x/merkledb/trie_test.go [hunk 1]
```diff
@@ -32,6 +46,10 @@ func getNodeValueWithBranchFactor(t ReadOnlyTrie, key string, bf BranchFactor) (
 		if err != nil {
 			return nil, err
 		}
+		if len(nodePath) == 0 {
```
Comment (danlaine) [inline x/merkledb/trie_test.go:49]: The comment on `getPathTo` says `// Always returns at least the root node.` so this should never happen, right?

File: x/merkledb/trieview.go [hunk 2]
```diff
@@ -246,10 +250,23 @@ func (t *trieView) calculateNodeIDs(ctx context.Context) error {
 			}
 		}
 
+		// If the root has no value and only a single child, that child can become the root.
+		// Use valueDigest instead of value because proof verification only sets the digest
```
Comment (danlaine) [inline x/merkledb/trieview.go:254]: If the root has no value and a single child, why is it `t.root` here? Shouldn't its child become the root in `remove`/`insert`?

File: x/merkledb/trieview.go [hunk 3]
```diff
@@ -737,12 +768,18 @@ func (t *trieView) deleteEmptyNodes(nodePath []*node) error {
...
-		// all node paths start at the root
-		currentNode      = t.root
-		matchedPathIndex = 0
-		nodes            = []*node{t.root}
+		// all paths start at the root
+		currentNode = t.root
+		nodes       []*node
 	)
 
+	if !key.HasPrefix(t.root.key) {
```
Comment (danlaine) [inline x/merkledb/trieview.go:776]:
```go
// Always returns at least the root node.
```
this is no longer true

**Review by danlaine — CHANGES_REQUESTED** [review]

File: x/merkledb/history.go [hunk 4]
```diff
@@ -53,9 +53,10 @@ type changeSummaryAndInsertNumber struct {
...
-	rootID ids.ID
-	nodes  map[Path]*change[*node]
-	values map[Path]*change[maybe.Maybe[[]byte]]
+	rootNode *node
+	rootID   ids.ID
+	nodes    map[Path]*change[*node]
+	values   map[Path]*change[maybe.Maybe[[]byte]]
 }
 
 func newChangeSummary(estimatedSize int) *changeSummary {
```
[the remark this answers is not in the document]
Comment (dboehm-avalabs) [reply x/merkledb/history.go:62]: I'm not sure what you mean.  They  are always populated once the changes are finalized.

File: x/merkledb/trieview.go (hunk 2 above)
Comment (dboehm-avalabs) [reply x/merkledb/trieview.go:254]: There is a tricky case where it was hard to detect in insert/remove.  It was easier to catch it here.

**Pull request closed without merging**